fsspec/implementations/http.py

```diff
@@ -262,9 +263,19 @@ async def _put_file(
         **kwargs,
     ):
         async def gen_chunks():
-            with open(rpath, "rb") as f:
-                callback.set_size(f.seek(0, 2))
-                f.seek(0)
+            # Support passing arbitrary file-like objects
+            # and use them instead of streams.
+            if isinstance(rpath, io.IOBase):
+                context = nullcontext(rpath)
```

Comment:
Although you have it the right way around, the signature of this method is wrong: rpath should be the remote one, and lpath the local one. We could fix this now.